RELEASE/scripts/autoscend/quests/level_12.ash

```diff
@@ -1429,79 +1433,47 @@ boolean L12_lastDitchFlyer()
 	{
 		return false;
 	}
-
-	auto_log_info("Not enough flyer ML but we are ready for the war... uh oh", "blue");
-
-	if(needStarKey())
-	{
-		if(!zone_isAvailable($location[The Hole in the Sky]))
-		{
-			return (L10_topFloor() || L10_holeInTheSkyUnlock());
-		}
-		else
-		{
-			if(LX_getStarKey())
-			{
-				return true;
-			}
-		}
-	}
-	else if(needDigitalKey())
+	if(my_level() < 13 && !isAboutToPowerlevel())
 	{
-		if(LX_getDigitalKey())
-		{
-			return true;
-		}
+		return false;		//let the powerlevel lock release first so we can do quests that are waiting for optimal conditions.
 	}
-	else
-	{
-		auto_log_warning("Should not have so little flyer ML at this point", "red");
-		wait(1);
-		if(!LX_attemptFlyering())
-		{
-			abort("Need more flyer ML but don't know where to go :(");
-		}
-		return true;
-	}
-	return false;
-}
 
-boolean LX_attemptFlyering()
-{
-	if(elementalPlanes_access($element[stench]) && auto_have_skill($skill[Summon Smithsness]))
-	{
-		return autoAdv(1, $location[Uncle Gator\'s Country Fun-Time Liquid Waste Sluice]);
-	}
-	else if(elementalPlanes_access($element[spooky]))
-	{
-		return autoAdv(1, $location[The Deep Dark Jungle]);
-	}
-	else if(elementalPlanes_access($element[cold]))
-	{
-		return autoAdv(1, $location[VYKEA]);
-	}
-	else if(elementalPlanes_access($element[stench]))
-	{
-		return autoAdv(1, $location[Uncle Gator\'s Country Fun-Time Liquid Waste Sluice]);
-	}
-	else if(elementalPlanes_access($element[sleaze]))
+	auto_log_info("Not enough flyer ML but we are ready for the war... uh oh", "blue");
+	if(LX_freeCombats(true)) return true;	//try to use free combats to make up the difference.
+
+	location scalezone = highestScalingZone();
+	float flyer_gains = my_buffedstat($stat[moxie]) + monster_level_adjustment();
+	switch(scalezone)
+	{
+		case $location[The Neverending Party]:
+			flyer_gains += 20; break;
+		case $location[VYKEA]:
+			flyer_gains += 6; break;
+		case $location[Uncle Gator\'s Country Fun-Time Liquid Waste Sluice]:
+		case $location[The Deep Dark Jungle]:
+		case $location[Sloppy Seconds Diner]:
+			flyer_gains += 5; break;
+	}
+	float adv_needed = (10000.0 - get_property("flyeredML").to_float()) / flyer_gains;
+	
+	warPlan plan_do_arena = auto_bestWarPlan();
+	plan_do_arena.do_arena = true;
+	warPlan plan_no_arena = auto_bestWarPlan();
+	plan_no_arena.do_arena = false;
+	int adv_saved = auto_warTotalBattles(plan_no_arena) - auto_warTotalBattles(plan_do_arena);
+	
+	if(adv_needed > adv_saved)
```

Comment:
comparing an int to a float. Not sure how ash handles that... should adv_needed be rounded before the compare?

Reply:
I don't think it should be rounded. you can not spend a fraction of an adventure. so if it is 7.01 then it needs 8 adventures. I originally though to ceil it but then realized there is no need. As `7.01 > 7 ` is `true`. thus the comparison is inherently ceiled.
on the other hand 6.9999 adv still takes 7 adv. so either way it should work out correctly this way

Reply:
Did you test those examples in mafia? Agree with your logic and should work on paper.

My main concern is the variables are different types. I've been burned by edge cases in the past where the programming language didn't compare them as I expected. Fix was to explicitly round/ceil/whatever so variable types were the same for the compare step

Reply:
I am actually concerned about the same thing. I twaked which was int and which was float before I got to a working version.
And I was specifically mixing int and float this way to avoid rounding issues. As comparing a "is float greater than int" seems to work well.
That said, it is an extreme edge case. and all that is at stake is 1 single adv being wasted.

Rounding would actually be harmful here making the answer wrong as well.
ceil would work but would be redundant as I would need to do to_int(ceil(float))

I did test it. However my discussion with you prompted me to do more testing. My code is working correctly. However I also tested replacing
`int adv_saved = auto_warTotalBattles(plan_no_arena) - auto_warTotalBattles(plan_do_arena);`
with
`float adv_saved = auto_warTotalBattles(plan_no_arena) - auto_warTotalBattles(plan_do_arena);`
and the results were identical.

as
```
> ash float test = 10+5; print(test == 15);

true

> ash float test = 10.0+5.0; print(test == 15);

true
```
As such, I am switching it to be all float to make the code more robust against future alterations.

Reply:
change made. it is now all floats